<a href="https://colab.research.google.com/github/BytesByBytes/BytezByBytez/blob/main/Lab_02_Literals%2C_Variables%2C_Expressions_and_Types.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Department of Computer Science and IT_ | **_La Trobe University, Australia_**

---

# Lab 02: Literals, Variables, Expressions, and Types
In this lab we will quickly familiarise ourselves with notebooks, then put into practice some of the programming concepts we've learnt by building a few programs.

## Notebooks Overview
Notebooks are interactive environments which allow you to write code and formatted text in the same place. As such, they are composed of two types of cells - text and code.

#### Text Cells
A text cell (like the one you're reading) is nothing more than plain text with some formatting applied. To edit the contents of a text cell, double-click the cell and make any changes, then press `<SHIFT>+<ENTER>` to apply the formatting. You won't be required to write any text cells, but it's good to know how they work. Try editing some text in this cell.

#### Code Cells
This type of cell contains code for whichever language the Notebook is configured - thus all of the notebooks in this subject support Python code. You can edit code cells in much the same way you would using any word processor. To run a cell's code, click somewhere in the cell and press `<CTRL>+<ENTER>`. Try modifying and running the below cell to see its outputs.

In [1]:
print('I am a code cell!')

I am a code cell!


## Programming Exercises

### Literals and Variables
Let's get coding. The next cell contains one variable and two literals - do you recall which is which? Variables have descriptive names like `home_city` or `patient_height` and contain values, whereas literals are just the values themselves - like `'Melbourne'` or `174.5`.

In the next cell replace `'Alice'` with your name as a string literal, then run the cell to see what's printed.

You should notice that the <ins>formatting of the</ins> output is slightly incorrect. Once you spot the problem, correct it and continue reading.

In [4]:
name = 'Stuart'
print('Your name is ' + name)

Your name is Stuart


###### Solution

The issue is that concatenation does exactly what you ask of it - it joins both strings as they're provided, and doesn't know anything about words being separated by spaces in English. For this reason, you will often need to add a space at the start or end of a string for correct formatting.

### Type Conversion

In the next cell we will try to print out a numerical data type - an integer. Although the code doesn't specify that the variable is an integer, Python infers the data type based on the fact that the initial value is a whole number.

Run the cell, and you'll encounter an error. What do you think the error means? Try to resolve it yourself before moving on.

In [6]:
lucky_number = 3
print('My lucky number is ' + str(lucky_number))

My lucky number is 3


###### Solution

The values on either side of the `+` operator above are different types - string and integer. Python doesn't allow adding strings with numerical data types, so an error is produced. \
To fix the code above, we need to convert `lucky_number` to a string. We do this by calling the `str` function - so replace `lucky_number` with `str(lucky_number)` on the last line above.

### More Type Conversion

What about other numerical data types? You'll often encounter the `float` data type, as this is capable of storing fractional numbers. Fix the code in the next cell using the `str` function twice only.

In [7]:
breed = 'labrador'
age = 3
height_cm = 57.5
print('My dog is a ' + breed + '. He is ' + str(age) + ' years old and ' + str(height_cm) + 'cm tall.')

My dog is a labrador. He is 3 years old and 57.5cm tall.


###### Solution

As `breed` is already a string, converting it using the `str` function won't have any effect. As such, only the numerical variables `age` and `height_cm` should be converted to a string when printing.

### Interactive Inputs and Data Types

None of the above code is interactive, meaning that the user can't input anything to the program, and it always outputs the same thing. We'll change that by using the `input` function. This captures the user's input as a string. \
Run the next cell and answer the questions when they are asked by the program.

In [ ]:
user_name = input('What is your name? ')
user_age = input('How old are you? ')
print('Your name is ' + user_name + ' and you are ' + user_age + ' years old.')

That program isn't very interesting, so now we'll extend it to provide more functionality. We want the program to tell us how old the user will be on their next birthday.

Add a line of code below which adds `1` to the user's age. Then, modify the print statement so it outputs an appropriate message.

In [12]:
user_name = input('What is your name? ')
user_age = int(input('How old are you? '))
user_age_next_birthday = user_age + 1 # Add 1 to the user's age here
print('Your name is ' + user_name + ' and you will be ' + str(user_age_next_birthday) + ' years old on your next birthday.')

What is your name? stuart
How old are you? 51
Your name is stuart and you will be 52 years old on your next birthday.


Uh oh - the program crashed! Do you know what the problem is? Try fixing it before continuing.

###### Solution

As mentioned earlier, the `input` function returns the user input as a string. So although you entered a number for the age question, `user_age` was of type `str`. The easiest way to fix this is to immediately convert it to an integer when taking the user's input. To do so, replace `input('How old are you? ')` with `int(input('How old are you? '))`.

It might look a little confusing, but just like in mathematics, parentheses indicate the order of operations. So the interpreter will take the user input, convert it to an `int`, then store it in `user_age`.

Once you've fixed it, run the code again - you'll realise that you need to convert the age back to a string again to print it.

You have now built your first interactive program - nice work!

### Choosing Data Types

Let's look at a slightly more complicated program. Read through the code below and see if you understand it all. Do you see any problems with it? Run the program and answer the program's prompts.

In [21]:
width_m = float(input('Width: '))
length_m = float(input('Length: '))
area_m2 = width_m * length_m
perimeter = 2 * width_m + 2 * length_m
print('Area: ' + str(area_m2) + ' sq. metres')
print('Perimeter: ' + str(perimeter) + ' metres')

Width: 2.5
Length: 2.5
Area: 6.25 sq. metres
Perimeter: 10.0 metres


This program calculates the area of a room using the equation $A = lw$, where $l$ and $w$ are the length and width, respectively. In order to verify that a program is working correctly, you need to think of some inputs to test it with, and know which outputs to expect. Try the below tests:

| Width | Length | Expected Area |
|-------|--------|---------------|
| 4     | 3      | 12            |
| 5     | 5      | 25            |
| 3.5   | 2      | 7             |
| 2.5   | 2.5    | 6.25          |

Although the program is mostly correct, you'll find that it only works if you enter whole numbers. Try again with fractional numbers - where and why does the program fail? Try to correct the error before continuing.

###### Solution

As we are explicitly converting the user inputs to integers, our program crashes if the user enters anything that's not a whole number. Depending on the purpose of the program, it might make sense to use either integer or floating point numbers - it's up to you as a programmer to make that decision!

In this case, it's not a very useful program if the user can only measure rooms in metre increments, so you should fix the bug by instead converting their inputs to `float`. Then, run the program again to ensure that it works correctly.

### Extending the Program

We'll now take the program one step further by also providing the user with the perimeter of the room. This is calculated with the equation $P = 2l + 2w$.

Update the program above to calculate the perimeter then display it with an additional print statement.

Manually calculate and complete the missing entries in the table below, then run your program to confirm that the perimeter is computed as expected.

| Width | Length | Expected Area | Expected Perimeter |
|-------|--------|---------------|--------------------|
| 4     | 3      | 12            | 14                 |
| 5     | 5      | 25            | 20                 |
| 3.5   | 2      | 7             | 11                 |
| 2.5   | 2.5    | 6.25          | 10                 |

_Double click this cell to edit the table, then press `<SHIFT>+<ENTER>` to apply the formatting._

### Constant Values

How about the perimeter of a circular room? The equation for this is $P = 2 \pi r$, where $r$ is the radius of the room.

As $\pi$ (pi) is a mathematical constant, it's best to store it in a variable of its own. To indicate that this is a constant that *never* changes, its variable name is all in UPPERCASE.

Write the rest of the program in the next cell to ask the user for the room's radius, then calculate and print out its perimeter. You should refer to the previous program if you get stuck.

We'll again use a table to verify our program, but as this is too difficult to calculate by hand, first complete the table with something trusted like a calculator.

| Radius | Expected Perimeter |
|--------|--------------------|
| 1      | 6.283186           |
| 0.5    | 3.141593           |
| 4      | 25.132744          |
| 2.5    | 15.707965          |

In [16]:
PI = 3.141593
radius = float(input('Radius Of Room:'))  # Ask the user for the radius of the room
perimeter = 2 * PI * radius  # Calculate the perimeter
area = PI * radius ** 2
print('Area: ' + str(area) + ' sq. metres')
print('Perimeter: ' + str(perimeter)) # Print the perimeter

Radius Of Room:2.5
Area: 19.63495625 sq. metres
Perimeter: 15.707965


The circular room program isn't quite as useful as the rectangular room program, as it doesn't calculate the area. Using the equation below, amend your program so that it also computes and prints the area.

$A = \pi r^2$

You should once again complete the missing fields in this table to verify your program.


| Radius | Expected Perimeter | Expected Area |
|--------|--------------------|---------------|
| 1      | 6.283186           | 3.141593      |
| 0.5    | 3.141593           | 0.78539825    |
| 4      | 25.132744          | 50.265488     |
| 2.5    | 15.707965          | 19.63495625   |


_Recall that the "to the power of" operator is `**`, so $x ^ 3$ is represented as `x ** 3` in Python._

## Bonus Tasks
If you have an appetite for more exercises, try solving the programming challenges below. It's highly recommended that you do so, as it will help get you "into the mind of a programmer".

For each of the exercises, take a moment to consider the below:
 * What are the inputs and outputs to this program?
 * What are the appropriate data types?
 * Ensure that you give the variables descriptive names (it's helpful to include the units if dealing with physical quantities like distance).
 * Test it with a variety of inputs to check that it is accurate.

### Velocity Calculator
The velocity of an object that takes $t$ seconds to travel $d$ metres is defined as $v = \frac{d}{t}$. \
Write a program that computes the velocity for the user.

In [18]:
# Write your velocity calculator solution here
time = float(input('Time: '))
distance = float(input('Distance: '))
velocity = distance / time
print('Velocity: ' + str(velocity))

Time: 2
Distance: 2
Velocity: 1.0


### Freefall Distance Calculator

The distance travelled by an object with constant acceleration $a$ for $t$ seconds is defined as $d = \frac{1}{2}at ^ 2$. \
Write a program that calculates the distance an object falls due to gravity, by using earth's _constant value_ acceleration of $a = 9.81$.

In [24]:
# Write your freefall distance calculator solution here
ACCELERATION = 9.81
time = float(input('Time in seconds: '))
distance = 0.5 * ACCELERATION * time ** 2
print('Distance travelled: ' + str(distance) + 'm')

Time in seconds: 12.21
Distance travelled: 731.2575105000002m
